README.md
# TheAnalyticsApi
Node / NPM client library for TheAnalyticsAPI.com

Easy to integrate Analytics, 

- Easily track events from your Node application
- Download the Aggregate totals & Insights
- Have the results pushed into your Datalake or anywhere else via Zapier or Slack with Webhooks

Currently in closed Alpha.

# Track an Event from your App

Basic example:

```javascript
const Analytics = require('@thatapicompany/theanalyticsapi');

const client = new Analytics('writeApiKey');

client.track({
  event: 'Signup'
});
```
n.b. Timestamp would be automatically added on the server.

Typescript / Babel:

```javascript
import Analytics  from '@thatapicompany/theanalyticsapi';

const client = new Analytics('writeApiKey');

client.track({
  event: 'Signup'
});
```

# For early access contact

Aden@ThatApiCompany.com

History.md

v1.0.6 / 2020-05-23
==========================

  * Fixed case issue with example code. Added Typescript basic example.

v1.0.4 / 2020-05-23
==========================

  * Updated optional timestamp sent to be in miliseconds

v1.0.2 / 2020-05-23
==========================

  * Released to NPM for Alpha testing


v1.0.0 / 2020-05-21
==========================

  * [Reference](https://github.com/segmentio/analytics-node/): Created Initial version using this library as basis


test.js
import { spy, stub } from 'sinon'
import bodyParser from 'body-parser'
import express from 'express'
import delay from 'delay'
import auth from 'basic-auth'
import pify from 'pify'
import test from 'ava'
import Analytics from '.'
import { version } from './package'

const noop = () => {}

const context = {
  library: {
    name: 'theanalyticsapi',
    version
  }
}

const metadata = { nodeVersion: process.versions.node }
const port = 4063

const createClient = options => {
  options = Object.assign({
    host: `http://localhost:${port}`
  }, options)

  const client = new Analytics('key', options)
  client.flush = pify(client.flush.bind(client))
  client.flushed = true

  return client
}

test.before.cb(t => {
  express()
    .use(bodyParser.json())
    .post('/api/track/batch', (req, res) => {
      const batch = req.body.batch

      const { name: writeKey } = auth(req)
      if (!writeKey) {
        return res.status(400).json({
          error: { message: 'missing write key' }
        })
      }

      const ua = req.headers['user-agent']
      if (ua !== `analytics-node/${version}`) {
        return res.status(400).json({
          error: { message: 'invalid user-agent' }
        })
      }

      if (batch[0] === 'error') {
        return res.status(400).json({
          error: { message: 'error' }
        })
      }

      if (batch[0] === 'timeout') {
        return setTimeout(() => res.end(), 5000)
      }

      res.json({})
    })
    .listen(port, t.end)
})

test('expose a constructor', t => {
  t.is(typeof Analytics, 'function')
})

test('require a write key', t => {
  t.throws(() => new Analytics(), 'You must pass your project\'s write key.')
})

test('create a queue', t => {
  const client = createClient()

  t.deepEqual(client.queue, [])
})

test('default options', t => {
  const client = new Analytics('key')

  t.is(client.writeKey, 'key')
  t.is(client.host, 'https://tanalytics-collector-api-6wyfjoyn3q-uc.a.run.app')
  t.is(client.flushAt, 20)
  t.is(client.flushInterval, 10000)
})

test('remove trailing slashes from `host`', t => {
  const client = new Analytics('key', { host: 'http://google.com///' })

  t.is(client.host, 'http://google.com')
})

test('overwrite defaults with options', t => {
  const client = new Analytics('key', {
    host: 'a',
    flushAt: 1,
    flushInterval: 2
  })

  t.is(client.host, 'a')
  t.is(client.flushAt, 1)
  t.is(client.flushInterval, 2)
})

test('keep the flushAt option above zero', t => {
  const client = createClient({ flushAt: 0 })

  t.is(client.flushAt, 1)
})

test('enqueue - add a message to the queue', t => {
  const client = createClient()

  const timestamp = new Date()
  client.enqueue('type', { timestamp }, noop)

  t.is(client.queue.length, 1)

  const item = client.queue.pop()

  t.is(typeof item.message.messageId, 'string')
  t.regex(item.message.messageId, /node-[a-zA-Z0-9]{32}/)
  t.deepEqual(item, {
    message: {
      timestamp,
      type: 'type',
      context,
      _metadata: metadata,
      messageId: item.message.messageId
    },
    callback: noop
  })
})

test('enqueue - stringify userId', t => {
  const client = createClient()

  client.track({
    userId: 10,
    event: 'event'
  }, noop)

  t.is(client.queue.length, 1)

  const item = client.queue.pop()

  t.is(item.message.anonymousId, undefined)
  t.is(item.message.userId, '10')
})
/*
test('enqueue - stringify anonymousId', t => {
  const client = createClient()

  client.screen({
    anonymousId: 157963456373623802,
    name: 'screen name'
  }, noop)

  t.is(client.queue.length, 1)

  const item = client.queue.pop()

  t.is(item.message.userId, undefined)
  // v8 will lose precision for big numbers.
  t.is(item.message.anonymousId, '157963456373623800')
})

test('enqueue - stringify ids handles strings', t => {
  const client = createClient()

  client.screen({
    anonymousId: '15796345',
    // We're explicitly testing the behaviour of the library if a customer
    // uses a String constructor.
    userId: new String('prateek'), // eslint-disable-line no-new-wrappers
    name: 'screen name'
  }, noop)

  t.is(client.queue.length, 1)

  const item = client.queue.pop()

  t.is(item.message.anonymousId, '15796345')
  t.is(item.message.userId.toString(), 'prateek')
})
*/
test('enqueue - don\'t modify the original message', t => {
  const client = createClient()
  const message = { event: 'test' }

  client.enqueue('type', message)

  t.deepEqual(message, { event: 'test' })
})

test('enqueue - flush on first message', t => {
  const client = createClient({ flushAt: 2 })
  client.flushed = false
  spy(client, 'flush')

  client.enqueue('type', {})
  t.true(client.flush.calledOnce)

  client.enqueue('type', {})
  t.true(client.flush.calledOnce)

  client.enqueue('type', {})
  t.true(client.flush.calledTwice)
})

test('enqueue - flush the queue if it hits the max length', t => {
  const client = createClient({
    flushAt: 1,
    flushInterval: null
  })

  stub(client, 'flush')

  client.enqueue('type', {})

  t.true(client.flush.calledOnce)
})

test('enqueue - flush after a period of time', async t => {
  const client = createClient({ flushInterval: 10 })
  stub(client, 'flush')

  client.enqueue('type', {})

  t.false(client.flush.called)
  await delay(20)

  t.true(client.flush.calledOnce)
})

test('enqueue - don\'t reset an existing timer', async t => {
  const client = createClient({ flushInterval: 10 })
  stub(client, 'flush')

  client.enqueue('type', {})
  await delay(5)
  client.enqueue('type', {})
  await delay(5)

  t.true(client.flush.calledOnce)
})

test('enqueue - extend context', t => {
  const client = createClient()

  client.enqueue('type', {
    event: 'test',
    context: { name: 'travis' }
  }, noop)

  const actualContext = client.queue[0].message.context
  const expectedContext = Object.assign({}, context, { name: 'travis' })

  t.deepEqual(actualContext, expectedContext)
})

test('enqueue - skip when client is disabled', async t => {
  const client = createClient({ enable: false })
  stub(client, 'flush')

  const callback = spy()
  client.enqueue('type', {}, callback)
  await delay(5)

  t.true(callback.calledOnce)
  t.false(client.flush.called)
})

test('flush - don\'t fail when queue is empty', async t => {
  const client = createClient()

  await t.notThrows(client.flush())
})
/*
test('flush - send messages', async t => {
  const client = createClient({ flushAt: 2 })

  const callbackA = spy()
  const callbackB = spy()
  const callbackC = spy()

  client.queue = [
    {
      message: 'a',
      callback: callbackA
    },
    {
      message: 'b',
      callback: callbackB
    },
    {
      message: 'c',
      callback: callbackC
    }
  ]

  const data = await client.flush()
  t.deepEqual(Object.keys(data), ['batch', 'timestamp', 'sentAt'])
  t.deepEqual(data.batch, ['a', 'b'])
  t.true(data.timestamp instanceof Date)
  t.true(data.sentAt instanceof Date)
  t.true(callbackA.calledOnce)
  t.true(callbackB.calledOnce)
  t.false(callbackC.called)
})
/*
test('flush - respond with an error', async t => {
  const client = createClient()
  const callback = spy()

  client.queue = [
    {
      message: 'error',
      callback
    }
  ]

  await t.throws(client.flush(), 'Bad Request')
})

test('flush - time out if configured', async t => {
  const client = createClient({ timeout: 500 })
  const callback = spy()

  client.queue = [
    {
      message: 'timeout',
      callback
    }
  ]

  await t.throws(client.flush(), 'timeout of 500ms exceeded')
})
*/
test('flush - skip when client is disabled', async t => {
  const client = createClient({ enable: false })
  const callback = spy()

  client.queue = [
    {
      message: 'test',
      callback
    }
  ]

  await client.flush()

  t.false(callback.called)
})
/*
test('identify - enqueue a message', t => {
  const client = createClient()
  stub(client, 'enqueue')

  const message = { userId: 'id' }
  client.identify(message, noop)

  t.true(client.enqueue.calledOnce)
  t.deepEqual(client.enqueue.firstCall.args, ['identify', message, noop])
})

test('identify - require a userId or anonymousId', t => {
  const client = createClient()
  stub(client, 'enqueue')

  t.throws(() => client.identify(), 'You must pass a message object.')
  t.throws(() => client.identify({}), 'You must pass either an "anonymousId" or a "userId".')
  t.notThrows(() => client.identify({ userId: 'id' }))
  t.notThrows(() => client.identify({ anonymousId: 'id' }))
})

test('group - enqueue a message', t => {
  const client = createClient()
  stub(client, 'enqueue')

  const message = {
    groupId: 'id',
    userId: 'id'
  }

  client.group(message, noop)

  t.true(client.enqueue.calledOnce)
  t.deepEqual(client.enqueue.firstCall.args, ['group', message, noop])
})

test('group - require a groupId and either userId or anonymousId', t => {
  const client = createClient()
  stub(client, 'enqueue')

  t.throws(() => client.group(), 'You must pass a message object.')
  t.throws(() => client.group({}), 'You must pass either an "anonymousId" or a "userId".')
  t.throws(() => client.group({ userId: 'id' }), 'You must pass a "groupId".')
  t.throws(() => client.group({ anonymousId: 'id' }), 'You must pass a "groupId".')
  t.notThrows(() => {
    client.group({
      groupId: 'id',
      userId: 'id'
    })
  })

  t.notThrows(() => {
    client.group({
      groupId: 'id',
      anonymousId: 'id'
    })
  })
})
*/
test('track - enqueue a message', t => {
  const client = createClient()
  stub(client, 'enqueue')

  const message = {
    userId: 1,
    event: 'event'
  }

  client.track(message, noop)

  t.true(client.enqueue.calledOnce)
  t.deepEqual(client.enqueue.firstCall.args, ['track', message, noop])
})

test('track - require event and either userId or anonymousId', t => {
  const client = createClient()
  stub(client, 'enqueue')

  t.throws(() => client.track(), 'You must pass a message object.')
  t.throws(() => client.track({}), 'You must pass either an "anonymousId" or a "userId".')
  t.throws(() => client.track({ userId: 'id' }), 'You must pass an "event".')
  t.throws(() => client.track({ anonymousId: 'id' }), 'You must pass an "event".')
  t.notThrows(() => {
    client.track({
      userId: 'id',
      event: 'event'
    })
  })

  t.notThrows(() => {
    client.track({
      anonymousId: 'id',
      event: 'event'
    })
  })
})
/*
test('page - enqueue a message', t => {
  const client = createClient()
  stub(client, 'enqueue')

  const message = { userId: 'id' }
  client.page(message, noop)

  t.true(client.enqueue.calledOnce)
  t.deepEqual(client.enqueue.firstCall.args, ['page', message, noop])
})

test('page - require either userId or anonymousId', t => {
  const client = createClient()
  stub(client, 'enqueue')

  t.throws(() => client.page(), 'You must pass a message object.')
  t.throws(() => client.page({}), 'You must pass either an "anonymousId" or a "userId".')
  t.notThrows(() => client.page({ userId: 'id' }))
  t.notThrows(() => client.page({ anonymousId: 'id' }))
})

test('screen - enqueue a message', t => {
  const client = createClient()
  stub(client, 'enqueue')

  const message = { userId: 'id' }
  client.screen(message, noop)

  t.true(client.enqueue.calledOnce)
  t.deepEqual(client.enqueue.firstCall.args, ['screen', message, noop])
})

test('screen - require either userId or anonymousId', t => {
  const client = createClient()
  stub(client, 'enqueue')

  t.throws(() => client.screen(), 'You must pass a message object.')
  t.throws(() => client.screen({}), 'You must pass either an "anonymousId" or a "userId".')
  t.notThrows(() => client.screen({ userId: 'id' }))
  t.notThrows(() => client.screen({ anonymousId: 'id' }))
})

test('alias - enqueue a message', t => {
  const client = createClient()
  stub(client, 'enqueue')

  const message = {
    userId: 'id',
    previousId: 'id'
  }

  client.alias(message, noop)

  t.true(client.enqueue.calledOnce)
  t.deepEqual(client.enqueue.firstCall.args, ['alias', message, noop])
})

test('alias - require previousId and userId', t => {
  const client = createClient()
  stub(client, 'enqueue')

  t.throws(() => client.alias(), 'You must pass a message object.')
  t.throws(() => client.alias({}), 'You must pass a "userId".')
  t.throws(() => client.alias({ userId: 'id' }), 'You must pass a "previousId".')
  t.notThrows(() => {
    client.alias({
      userId: 'id',
      previousId: 'id'
    })
  })
})
*/
test('isErrorRetryable', t => {
  const client = createClient()

  t.false(client._isErrorRetryable({}))

  // ETIMEDOUT is retryable as per `is-retry-allowed` (used by axios-retry in `isNetworkError`).
  t.true(client._isErrorRetryable({ code: 'ETIMEDOUT' }))

  // ECONNABORTED is not retryable as per `is-retry-allowed` (used by axios-retry in `isNetworkError`).
  t.false(client._isErrorRetryable({ code: 'ECONNABORTED' }))

  t.true(client._isErrorRetryable({ response: { status: 500 } }))
  t.true(client._isErrorRetryable({ response: { status: 429 } }))

  t.false(client._isErrorRetryable({ response: { status: 200 } }))
})
/*
test('allows messages > 32kb', t => {
  const client = createClient()

  const event = {
    userId: 1,
    event: 'event',
    properties: {}
  }
  for (var i = 0; i < 10000; i++) {
    event.properties[i] = 'a'
  }

  t.notThrows(() => {
    client.track(event, noop)
  })
})
*/


cli.js
#!/usr/bin/env node
'use strict'

const program = require('commander')
const Analytics = require('.')
const pkg = require('./package')

const toObject = str => JSON.parse(str)

program
  .version(pkg.version)
  .option('-w, --writeKey <key>', 'the Segment write key to use')
  .option('-h, --host <host>', 'the Segment API hostname to use')
  .option('-t, --type <type>', 'the Segment message type')

  .option('-u, --userId <id>', 'the user id to send the event as')
  .option('-a, --anonymousId <id>', 'the anonymous user id to send the event as')
  .option('-c, --context <context>', 'additional context for the event (JSON-encoded)', toObject)
  .option('-i, --integrations <integrations>', 'additional integrations for the event (JSON-encoded)', toObject)

  .option('-e, --event <event>', 'the event name to send with the event')
  .option('-p, --properties <properties>', 'the event properties to send (JSON-encoded)', toObject)

  .option('-n, --name <name>', 'name of the screen or page to send with the message')
  .option('-t, --traits <traits>', 'the identify/group traits to send (JSON-encoded)', toObject)
  .option('-g, --groupId <groupId>', 'the group id')
  .option('-pid, --previousId <previousId>', 'the previous id')

  .parse(process.argv)

if (program.args.length !== 0) {
  program.help()
}

const writeKey = program.writeKey
const host = program.host
const type = program.type

const userId = program.userId
const anonymousId = program.anonymousId
const context = program.context
const integrations = program.integrations

const event = program.event
const properties = program.properties
const name = program.name
const traits = program.traits
const groupId = program.groupId
const previousId = program.previousId

const run = (method, args) => {
  const analytics = new Analytics(writeKey, { host, flushAt: 1 })
  analytics[method](args, err => {
    if (err) {
      console.error(err.stack)
      process.exit(1)
    }
  })
}

switch (type) {
  case 'track':
    run('track', {
      event,
      properties,
      userId,
      anonymousId,
      context,
      integrations
    })
    break
  default:
    console.error('invalid type:', type)
    process.exit(1)
}


index.js
'use strict'

const assert = require('assert')
const removeSlash = require('remove-trailing-slash')
const axios = require('axios')
const axiosRetry = require('axios-retry')
const ms = require('ms')
const uuid = require('uuid/v4')
const md5 = require('md5')
const version = require('./package.json').version
const isString = require('lodash.isstring')

// replace
const looselyValidate = require('@segment/loosely-validate-event')

const setImmediate = global.setImmediate || process.nextTick.bind(process)
const noop = () => {}

class TheAnalyticsAPI {
  /**
   * Initialize a new `Analytics` with your Segment project's `writeKey` and an
   * optional dictionary of `options`.
   *
   * @param {String} writeKey
   * @param {Object} [options] (optional)
   *   @property {Number} flushAt (default: 20)
   *   @property {Number} flushInterval (default: 10000)
   *   @property {String} host (default: 'https://api.segment.io')
   *   @property {Boolean} enable (default: true)
   */

  constructor (writeKey, options) {
    options = options || {}

    assert(writeKey, 'You must pass your project\'s write key.')
    this.queue = []
    this.writeKey = writeKey
    this.host = removeSlash(options.host || 'https://tanalytics-collector-api-6wyfjoyn3q-uc.a.run.app')
    this.timeout = options.timeout || false
    this.flushAt = Math.max(options.flushAt, 1) || 20
    this.flushInterval = options.flushInterval || 10000
    this.flushed = false
    Object.defineProperty(this, 'enable', {
      configurable: false,
      writable: false,
      enumerable: true,
      value: typeof options.enable === 'boolean' ? options.enable : true
    })

    axiosRetry(axios, {
      retries: options.retryCount || 3,
      retryCondition: this._isErrorRetryable,
      retryDelay: axiosRetry.exponentialDelay
    })
  }

  _validate (message, type) {
    try {
      looselyValidate(message, type)
    } catch (e) {
      throw e
    }
  }

  /**
   * Send a track `message`.
   *
   * @param {Object} message
   * @param {Function} [callback] (optional)
   * @return {TheAnalyticsAPI}
   */

  track (message, callback) {
    this._validate(message, 'track')
    this.enqueue('track', message, callback)
    return this
  }

  /**
   * Add a `message` of type `type` to the queue and
   * check whether it should be flushed.
   *
   * @param {String} type
   * @param {Object} message
   * @param {Function} [callback] (optional)
   * @api private
   */

  enqueue (type, message, callback) {
    callback = callback || noop

    if (!this.enable) {
      return setImmediate(callback)
    }

    message = Object.assign({}, message)
    message.type = type
    message.context = Object.assign({
      library: {
        name: 'theanalyticsapi',
        version
      }
    }, message.context)

    message._metadata = Object.assign({
      nodeVersion: process.versions.node
    }, message._metadata)

    if (!message.timestamp) {
      message.timestamp = new Date().getTime();
    }

    if (!message.messageId) {
      // We md5 the messaage to add more randomness. This is primarily meant
      // for use in the browser where the uuid package falls back to Math.random()
      // which is not a great source of randomness.
      // Borrowed from analytics.js (https://github.com/segment-integrations/analytics.js-integration-segmentio/blob/a20d2a2d222aeb3ab2a8c7e72280f1df2618440e/lib/index.js#L255-L256).
      message.messageId = `node-${md5(JSON.stringify(message))}-${uuid()}`
    }

    if (message.anonymousId && !isString(message.anonymousId)) {
      message.anonymousId = JSON.stringify(message.anonymousId)
    }
    if (message.userId && !isString(message.userId)) {
      message.userId = JSON.stringify(message.userId)
    }

    this.queue.push({ message, callback })

    if (!this.flushed) {
      this.flushed = true
      this.flush()
      return
    }

    if (this.queue.length >= this.flushAt) {
      this.flush()
    }

    if (this.flushInterval && !this.timer) {
      this.timer = setTimeout(this.flush.bind(this), this.flushInterval)
    }
  }

  /**
   * Flush the current queue
   *
   * @param {Function} [callback] (optional)
   * @return {Analytics}
   */

  flush (callback) {
    callback = callback || noop

    if (!this.enable) {
      return setImmediate(callback)
    }

    if (this.timer) {
      clearTimeout(this.timer)
      this.timer = null
    }

    if (!this.queue.length) {
      return setImmediate(callback)
    }

    const items = this.queue.splice(0, this.flushAt)
    const callbacks = items.map(item => item.callback)
    const messages = items.map(item => item.message)

    const data = {
      batch: messages,
      timestamp: new Date().getTime(),
      sentAt: new Date().getTime()
    }

    const done = err => {
      callbacks.forEach(callback => callback(err))
      callback(err, data)
    }

    const headers = {}

    headers['user-agent'] = `theanalyticsapi-client-node/${version}`
    headers['api_key'] = this.writeKey

    const req = {
      method: 'POST',
      url: `${this.host}/api/track/batch`,
      /* auth: {
        username: this.writeKey
      }, */
      data,
      headers
    }

    if (this.timeout) {
      req.timeout = typeof this.timeout === 'string' ? ms(this.timeout) : this.timeout
    }

    axios(req)
      .then(() => done())
      .catch(err => {
        if (err.response) {
          const error = new Error(err.response.statusText)
          return done(error)
        }

        done(err)
      })
  }

  _isErrorRetryable (error) {
    // Retry Network Errors.
    if (axiosRetry.isNetworkError(error)) {
      return true
    }

    if (!error.response) {
      // Cannot determine if the request can be retried
      return false
    }

    // Retry Server Errors (5xx).
    if (error.response.status >= 500 && error.response.status <= 599) {
      return true
    }

    // Retry if rate limited.
    if (error.response.status === 429) {
      return true
    }

    return false
  }
}

module.exports = TheAnalyticsAPI
